wp-includes/mysql/class-wp-mysql-lexer.php

```diff
@@ -0,0 +1,2898 @@
+	 * version, and how some tokens should be interpreted.
+	 *
+	 * @var int
+	 */
+	private $mysql_version;
+
+	/**
+	 * The SQL modes that should be considered active during tokenization.
+	 *
+	 * This is an integer that represents currently active SQL modes as a bitmask.
+	 * The SQL modes are defined as "SQL_MODE_"-prefixed constants in this class.
+	 * The list of the SQL modes isn't exhaustive, as only some affect tokenization.
+	 *
+	 * @var int
+	 */
+	private $sql_modes;
+
+	/**
+	 * How many bytes from the original SQL payload have been read and tokenized.
+	 *
+	 * This is an internal cursor that is used to track the current position in
+	 * the SQL payload during tokenization. When used as an index in the SQL
+	 * payload, it points to the next byte to read.
+	 *
+	 * @var int
+	 */
+	private $bytes_already_read = 0;
+
+	/**
+	 * Byte offset in the SQL payload where current token starts.
+	 *
+	 * This is used to extract the token bytes after the token is processed.
+	 * The bytes of the current token are represented by "$this->sql" in range
+	 * from "$this->token_starts_at" to "$this->bytes_already_read - 1".
+	 *
+	 * @var int
+	 */
+	private $token_starts_at = 0;
+
+	/**
+	 * Whether the tokenizer is inside an active MySQL-specific comment.
+	 *
+	 * MySQL supports a special comment syntax whose content is recognized as
+	 * a comment by most database engines, but can be treated as SQL by MySQL:
+	 *
+	 *  1. /*! ...  - The content is treated as SQL.
+	 *  2. /*!12345 - The content is treated as SQL when "MySQL version >= 12345".
+	 *
+	 * @var bool
+	 */
+	private $in_mysql_comment = false;
+
+	/**
+	 * @param string $sql The SQL payload to tokenize.
+	 * @param int $mysql_version The version of the MySQL server that the SQL payload is intended for.
+	 * @param int $sql_modes The SQL modes that should be considered active during tokenization.
+	 */
+	public function __construct(
+		string $sql,
+		int $mysql_version = 80038,
+		int $sql_modes = 0
+	) {
+		$this->sql           = $sql;
+		$this->mysql_version = $mysql_version;
+		$this->sql_modes     = $sql_modes;
+	}
+
+	/**
+	 * Read the next token from the SQL payload and return it as a token object.
+	 *
+	 * This method reads bytes from the SQL payload until a token is recognized.
+	 * It starts from "$this->sql[ $this->bytes_already_read ]", advances the
+	 * number of bytes read, and returns a WP_MySQL_Token object. When the end of
+	 * the SQL payload is reached, the method always returns an EOF token.
+	 *
+	 * @return WP_MySQL_Token A token object representing the next recognized token.
+	 */
+	public function next_token(): WP_MySQL_Token {
+		do {
+			$this->token_starts_at = $this->bytes_already_read;
+			$type                  = $this->read_next_token();
+		} while (
+			in_array(
+				$type,
+				array(
+					self::WHITESPACE,
+					self::COMMENT,
+					self::MYSQL_COMMENT_START,
+					self::MYSQL_COMMENT_END,
+				),
+				true
+			)
+		);
```

Comment:
Style nitpick – a direct comparison seems more readable here:

```suggestion
		} while (
			self::WHITESPACE === $type ||
			self::COMMENT === $type ||
			self::MYSQL_COMMENT_START === $type ||
			self::MYSQL_COMMENT_END === $type
		);
```

Reply:
@adamziel True. The interesting thing here is that `in_array` is slightly faster. It goes from around 1.05s to 1.10s for all the 70k queries on a powerful MacBook Pro. In the whole parsing pipeline, it's nothing significant, of course, so I think it's OK to change that. What do you think?

Reply:
Oh, how interesting! It seems like an okay price to pay for more readability in one of the most important loops.

Reply:
@adamziel Agreed. This may still evolve in the future, e.g. if the skipped tokens are configurable.